### PR TITLE
record: Refactoring script functions of record time

### PR DESCRIPTION
**Pull request by Taeung**
To clean up `mcount_{entry,exit}_filter_record()` that
seems too big, make script processing code in them
new functions `script_hook_{entry,exit}()`.

I tested this PR like below:
```
$ ./runtest.py script
Test case                 pg             finstrument-fu
------------------------: O0 O1 O2 O3 Os O0 O1 O2 O3 Os
157 script_python       : OK OK OK OK OK OK OK OK OK OK
169 script_args         : OK OK OK OK OK OK OK OK OK OK
170 script_filter       : OK OK OK OK OK OK OK OK OK OK
171 script_option       : OK OK OK OK OK OK OK OK OK OK
```

### CODE REVIEW

File: libmcount/mcount.c [hunk 1]
```diff
@@ -585,6 +591,45 @@ void mcount_entry_filter_record(struct mcount_thread_data *mtdp,
 
 }
 
+static void script_hook_exit(struct mcount_thread_data *mtdp,
+			     struct mcount_ret_stack *rstack)
```
Comment (namhyung) [inline libmcount/mcount.c]: Could you please move it close to the `script_hook_entry`?

Comment (Taeung) [reply libmcount/mcount.c]: Yep !

**Comment by Taeung** [comment]
@namhyung , I moved to `script_hook_exit()` to the `script_hook_entry()`.

However, I thought new `script_save_context()` because of duplicated code
of script_hook_{entry,exit}().
what about the below patch ?

```patch
diff --git a/libmcount/mcount.c b/libmcount/mcount.c
index 1ccccf7..e9537f8 100644
--- a/libmcount/mcount.c
+++ b/libmcount/mcount.c
@@ -449,6 +449,83 @@ enum filter_result mcount_entry_filter_check(struct mcount_thread_data *mtdp,
 	return FILTER_IN;
 }
 
+static int script_save_context(struct script_context *sc_ctx,
+			       struct mcount_thread_data *mtdp,
+			       struct mcount_ret_stack *rstack,
+			       struct sym *sym, bool has_arg_retval,
+			       struct list_head *pargs)
+{
+	unsigned long entry_addr = rstack->child_ip;
+	char *symname = symbol_getname(sym, entry_addr);
+
+	if (!script_match_filter(symname)) {
+		symbol_putname(sym, symname);
+		return -1;
+	}
+
+	sc_ctx->tid       = mcount_gettid(mtdp);
+	sc_ctx->depth     = rstack->depth;
+	sc_ctx->address   = entry_addr;
+	sc_ctx->name      = symname;
+	sc_ctx->timestamp = rstack->start_time;
+	if (rstack->end_time)
+		sc_ctx->duration =
+			rstack->end_time - rstack->start_time;
+
+	if (has_arg_retval) {
+		unsigned *argbuf = get_argbuf(mtdp, rstack);
+
+		sc_ctx->arglen  = argbuf[0];
+		sc_ctx->argbuf  = &argbuf[1];
+		sc_ctx->argspec = pargs;
+	}
+	else {
+		/* prevent access to arguments */
+		sc_ctx->arglen  = 0;
+	}
+
+	return 0;
+}
+
+static void script_hook_entry(struct mcount_thread_data *mtdp,
+			      struct mcount_ret_stack *rstack,
+			      struct uftrace_trigger *tr)
+{
+	struct script_context sc_ctx;
+	struct sym *sym = find_symtabs(&symtabs, rstack->child_ip);
+
+	if (script_save_context(&sc_ctx, mtdp, rstack, sym,
+				tr->flags & TRIGGER_FL_ARGUMENT,
+				tr->pargs) < 0)
+		return;
+
+	/* accessing argument in script might change arch-context */
+	mcount_save_arch_context(&mtdp->arch);
+	script_uftrace_entry(&sc_ctx);
+	mcount_restore_arch_context(&mtdp->arch);
+
+	symbol_putname(sym, sc_ctx.name);
+}
+
+static void script_hook_exit(struct mcount_thread_data *mtdp,
+			     struct mcount_ret_stack *rstack)
+{
+	struct script_context sc_ctx;
+	struct sym *sym = find_symtabs(&symtabs, rstack->child_ip);
+
+	if (script_save_context(&sc_ctx, mtdp, rstack, sym,
+				rstack->flags & MCOUNT_FL_RETVAL,
+				rstack->pargs) < 0)
+		return;
+
+	/* accessing argument in script might change arch-context */
+	mcount_save_arch_context(&mtdp->arch);
+	script_uftrace_exit(&sc_ctx);
+	mcount_restore_arch_context(&mtdp->arch);
+
+	symbol_putname(sym, sc_ctx.name);
+}
+
 /* save current filter state to rstack */
 void mcount_entry_filter_record(struct mcount_thread_data *mtdp,
 				struct mcount_ret_stack *rstack,
@@ -532,41 +609,8 @@ void mcount_entry_filter_record(struct mcount_thread_data *mtdp,
 		}
 
 		/* script hooking for function entry */
-		if (SCRIPT_ENABLED && script_str) {
-			unsigned long entry_addr = rstack->child_ip;
-			struct sym *sym = find_symtabs(&symtabs, entry_addr);
-			char *symname = symbol_getname(sym, entry_addr);
-			struct script_context sc_ctx;
-
-			if (!script_match_filter(symname))
-				goto skip;
-
-			sc_ctx.tid       = mcount_gettid(mtdp);
-			sc_ctx.depth     = rstack->depth;
-			sc_ctx.timestamp = rstack->start_time;
-			sc_ctx.address   = entry_addr;
-			sc_ctx.name      = symname;
-
-			if (tr->flags & TRIGGER_FL_ARGUMENT) {
-				unsigned *argbuf = get_argbuf(mtdp, rstack);
-
-				sc_ctx.arglen  = argbuf[0];
-				sc_ctx.argbuf  = &argbuf[1];
-				sc_ctx.argspec = tr->pargs;
-			}
-			else {
-				/* prevent access to arguments */
-				sc_ctx.arglen  = 0;
-			}
-
-			/* accessing argument in script might change arch-context */
-			mcount_save_arch_context(&mtdp->arch);
-			script_uftrace_entry(&sc_ctx);
-			mcount_restore_arch_context(&mtdp->arch);
-
-skip:
-			symbol_putname(sym, symname);
-		}
+		if (SCRIPT_ENABLED && script_str)
+			script_hook_entry(mtdp, rstack, tr);
 
 #define FLAGS_TO_CHECK  (TRIGGER_FL_RECOVER | TRIGGER_FL_TRACE_ON | TRIGGER_FL_TRACE_OFF)
 
@@ -657,42 +701,8 @@ void mcount_exit_filter_record(struct mcount_thread_data *mtdp,
 		}
 
 		/* script hooking for function exit */
-		if (SCRIPT_ENABLED && script_str) {
-			unsigned long entry_addr = rstack->child_ip;
-			struct sym *sym = find_symtabs(&symtabs, entry_addr);
-			char *symname = symbol_getname(sym, entry_addr);
-			struct script_context sc_ctx;
-
-			if (!script_match_filter(symname))
-				goto skip;
-
-			sc_ctx.tid       = mcount_gettid(mtdp);
-			sc_ctx.depth     = rstack->depth;
-			sc_ctx.timestamp = rstack->start_time;
-			sc_ctx.duration  = rstack->end_time - rstack->start_time;
-			sc_ctx.address   = entry_addr;
-			sc_ctx.name      = symname;
-
-			if (rstack->flags & MCOUNT_FL_RETVAL) {
-				unsigned *argbuf = get_argbuf(mtdp, rstack);
-
-				sc_ctx.arglen  = argbuf[0];
-				sc_ctx.argbuf  = &argbuf[1];
-				sc_ctx.argspec = rstack->pargs;
-			}
-			else {
-				/* prevent access to retval*/
-				sc_ctx.arglen  = 0;
-			}
-
-			/* accessing argument in script might change arch-context */
-			mcount_save_arch_context(&mtdp->arch);
-			script_uftrace_exit(&sc_ctx);
-			mcount_restore_arch_context(&mtdp->arch);
-
-skip:
-			symbol_putname(sym, symname);
-		}
+		if (SCRIPT_ENABLED && script_str)
+			script_hook_exit(mtdp, rstack);
 	}
 }
```

**Comment by namhyung** [comment]
Looks good.  And I'd like to leave it as a function for now.  But I think it's good to have the getname/putname pair in a same function.

**Comment by Taeung** [comment]
Yep, I got it, will fix it.

**Comment by Taeung** [comment]
I changed this PR based on your comments.

v3:
 - Use the getname/putname pair in a same function.
 - Add `script_save_context()`

File: libmcount/mcount.c [hunk 2]
```diff
@@ -449,6 +449,85 @@ enum filter_result mcount_entry_filter_check(struct mcount_thread_data *mtdp,
 	return FILTER_IN;
 }
 
+static int script_save_context(struct script_context *sc_ctx,
+			       struct mcount_thread_data *mtdp,
+			       struct mcount_ret_stack *rstack,
+			       struct sym *sym, char *symname,
```
Comment (namhyung) [inline libmcount/mcount.c]: I think the `sym` parameter can be omitted.

File: libmcount/mcount.c [hunk 3]
```diff
@@ -449,6 +449,85 @@ enum filter_result mcount_entry_filter_check(struct mcount_thread_data *mtdp,
...
+		return -1;
+
+	sc_ctx->tid       = mcount_gettid(mtdp);
+	sc_ctx->depth     = rstack->depth;
+	sc_ctx->address   = rstack->child_ip;
+	sc_ctx->name      = symname;
+	sc_ctx->timestamp = rstack->start_time;
+	if (rstack->end_time)
+		sc_ctx->duration =
+			rstack->end_time - rstack->start_time;
```
Comment (namhyung) [inline libmcount/mcount.c]: Just a style nit: please keep it in the same line.

File: libmcount/mcount.c (hunk 2 above)
Comment (Taeung) [reply libmcount/mcount.c]: Oops.. will fix it

**Comment by Taeung** [comment]
@namhyung , I resent PR based on your comments 😃

**Pull request merged**